### PR TITLE
Report aria-description always

### PR DESCRIPTION
**Pull request by feerrenrut**
**Reviewers - There are several commits to refactor / fix issues encountered. It will be easier to review the changes one commit at a time. I have already grouped and squashed related changes. Commits with messages starting with "Fixup / squash" will be squashed into the commit they reference**

### Link to issue number:
None

### Summary of the issue:
The HTML attribute `aria-description` has been introduced as one aspect of supporting a broader concept of annotations on the web. It maps to `accDescription`. In contrast to other sources for `accDescription` such as `title`, our expectation is annotations should be reported whenever present.

### Description of how this pull request fixes the issue:
Experimental support has been added for reporting aria-description.

### Testing strategy:
System tests for Chrome checking speech output in browse mode, focus mode, and say-all.
- The feature for `description-from` is only available in Chrome ` 92.0.4479.0+`
- The [most recent Appveyor has is Chrome `90.0.4430.72`](https://www.appveyor.com/docs/windows-images-software/#web-browsers)
- For now the assert for reporting description is expected to fail.

Manual testing with Braille, mouse, and object navigation.

### Known issues with pull request:
- Firefox and Edge do not yet support aria-description as expected.
- Reporting due to mouse seems broken in Chrome?
- [x] Garbage handler warnings: Fixed with #12617

### Change log entries:
New features:
`aria-description will now be reported by default.`

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Comment by feerrenrut** [comment]
I suspect that the version of chrome on appveyor does not yet support the "descriptionFrom" attribute.

**Comment by CyrilleB79** [comment]
Why is the option named "Report aria-description always" and not just "Report aria-description"?

**Comment by feerrenrut** [comment]
> Why is the option named "Report aria-description always" and not just "Report aria-description"?

In the past we reported `accDescription` under certain (but not all) circumstances. `aria-description` is one source for `accDescription`. We'll continue to report `accDescription` under the same circumstances as before, unless the source is `aria-description` in which case we'll always report it. Turning it off does not mean `aria-description` will never be reported.

As an advanced setting the naming is developer centric. Happy to hear suggestions.

File: source/NVDAObjects/JAB/__init__.py
Comment (seanbudd) [inline source/NVDAObjects/JAB/__init__.py:2]: This is missing serval contributors from git blame, and I think "Mozilla Corporation" as a copy/paste mistake. However, now that the message is added, the list of contributors can be updated.

```suggestion
# Copyright (C) 2007-2021 NV Access Limited
```

File: source/brailleHelpers.py
Comment (seanbudd) [inline source/brailleHelpers.py:18]:
```suggestion
	""" Determines if the given description should be reported. Returns None if description should not be
	reported, otherwise returns the description to report.
	"""
```

File: source/compoundDocuments.py
Comment (seanbudd) [inline source/compoundDocuments.py]: This is duplicated in the `else` condition

File: source/speech/speech.py
Comment (seanbudd) [inline source/speech/speech.py]: This should have been a review comment - `alwaysReportDescription` isn't set anywhere or respected by braille or used when aria-description is the source. I'm not sure if this is an issue.

```suggestion
				or attrs.get('alwaysReportDescription', False)
```

File: source/speech/speech.py
Comment (seanbudd) [inline source/speech/speech.py]: Note that we use `if description is not None` further down to generate the `descriptionSequence`

File: source/speechXml.py
Comment (seanbudd) [inline source/speechXml.py:54]: An optional change, that needs documenting

**Review by seanbudd — CHANGES_REQUESTED** [review]

File: source/brailleHelpers.py
Comment (feerrenrut) [reply source/brailleHelpers.py:18]: I'm going to drop this helper function. DRY is good, but this is only used twice, and those usages haven't been around long, we also don't want hasty abstractions.

File: source/compoundDocuments.py
Comment (feerrenrut) [reply source/compoundDocuments.py]: I think the initial intention was to document the various types that had to be handled. I'll do this in a more concise way.

File: source/brailleHelpers.py
Comment (seanbudd) [reply source/brailleHelpers.py:18]: This makes sense. I was hoping to reduce the complexity of the functions which use it, and generally `braille.py` further (or at least prevent from adding to it). But that can wait for a more focused change on `braille.py`.

File: source/speechXml.py
Comment (feerrenrut) [reply source/speechXml.py:54]: I think I'll leave this. Given the context of how this is used I think it is ok for it to be part of the public API for speechXML

**Comment by CyrilleB79** [comment]
> > Why is the option named "Report aria-description always" and not just "Report aria-description"?
> 
> In the past we reported `accDescription` under certain (but not all) circumstances. `aria-description` is one source for `accDescription`. We'll continue to report `accDescription` under the same circumstances as before, unless the source is `aria-description` in which case we'll always report it. Turning it off does not mean `aria-description` will never be reported.
> 
> As an advanced setting the naming is developer centric. Happy to hear suggestions.

I am a developper and tester but I do not know the whole codebase of NVDA. More specifically, I did not know that only some descriptions were reported, not all (provided the corresponding option in 'Object presentation' is checked).

Regarding a suggestion to rephrase this option, it seems hard to have a one-line formulation. I would suggest to add more details in the user guide.
There could be a dedicated paragraph with more details for this option in the user guide rather than a paragraph for the whole ARIA group of options. E.g.:

> - "Report aria-description always": enables reporting of aria-descriptions.
> If this option is checked, aria-description is always reported
> If this option is unchecked, legacy pre-2021.2 behaviour occurs. (please describe more precisely the legacy behaviour)
> Note: aria-description only applies to the web, and is only a subset of the sources for "description".

**Comment by feerrenrut** [comment]
I can add more details to the userguide. But do note these features and options are unlikely to stay the same. This option exists so that NV Access developers working on this can iteratively progress the related features. It also allows us to give specific instructions to browser developers to enable this for testing.

File: user_docs/en/userGuide.t2t
Comment (CyrilleB79) [inline user_docs/en/userGuide.t2t:1858]: Seems good to me. Whoever is interested in knowing what this option does will have the key words to search for more information if needed.

File: source/virtualBuffers/webKit.py
Comment (lukaszgo1) [inline source/virtualBuffers/webKit.py:21]: With what WebKit controls  this has been tested? The only one known to me is Itunes Store controls in Itunes and I find it hard to believe that `aria-description` is used there. If the intent was to provide different implementations for Chromium based browsers and Firefox this code should be placed in `NVDAObjects.IAccessible.chromium.ChromeVBuf` instead.

**Comment by feerrenrut** [comment]
Blocked by #12617

File: source/virtualBuffers/webKit.py
Comment (feerrenrut) [reply source/virtualBuffers/webKit.py]: This has been addressed.

**Review by feerrenrut — COMMENTED** [review]
I have done a self review.

**Comment by feerrenrut** [comment]
I have rebased this to condense all the "fixup" commits, and merge in the latest from master. When approved this can be merged as a squash merge. Keeping the commits separated is useful for reviewing this PR, but not entirely required once merged to master. There is no moved code, only modifications.

**Review by michaelDCurran — APPROVED** [review]
I skipped the first 4 commits and started reviewing at "Remove ignoreEditableText param...".
 The rest looks good to me.

File: source/NVDAObjects/IAccessible/mozilla.py
Comment (feerrenrut) [reply source/NVDAObjects/IAccessible/mozilla.py:33]: Ah ok. Thanks. I'll update this.

File: tests/system/libraries/AssertsLib.py
Comment (feerrenrut) [inline tests/system/libraries/AssertsLib.py:16]: Add comment: 'Include text in robot test report so that actual behavior can be determined (using the report) even when the test passes.'

**Pull request merged**
